File: fixtures/p1/sample.rs
use std::sync::Arc;

pub trait UserRepository {
    fn find_by_id(&self, id: &str) -> Result<User, RepoError>;
    fn save(&self, user: User) -> Result<(), RepoError>;
}

pub struct UserService {
    repo: Arc<dyn UserRepository>,
    logger: Logger,
}

impl UserService {
    pub fn new(repo: Arc<dyn UserRepository>, logger: Logger) -> Self {
        Self { repo, logger }
    }

    pub async fn get_user(&self, id: &str) -> Result<User, ServiceError> {
        todo!()
    }
}

File: fixtures/p1/sample.py
from abc import ABC, abstractmethod
from typing import Optional, Protocol


class UserRepository(Protocol):
    """Loads users from persistence."""

    def find_by_id(self, user_id: str) -> "User":
        ...


class AuditSink(ABC):
    @abstractmethod
    def record(self, event: str) -> None:
        ...


class UserService:
    repository: UserRepository

    def __init__(
        self,
        repository: UserRepository,
        audit: Optional[AuditSink] = None,
    ) -> None:
        self.repository = repository
        self.audit = audit

    @property
    def enabled(self) -> bool:
        return True

    @enabled.setter
    def enabled(self, value: bool) -> None:
        pass

    async def get_user(self, user_id: str) -> "User":
        return self.repository.find_by_id(user_id)

    def _reset_cache(self) -> None:
        pass


async def build_user_service(repository: UserRepository) -> UserService:
    return UserService(repository)
